health/health.go

```diff
@@ -0,0 +1,50 @@
+package health
+
+import (
+	"context"
+	"errors"
+	"net/http"
+)
+
+type HealthCheck interface {
+	RegisterHealthCheck(fn HealthCheckFunc)
+	HTTPHandler() http.HandlerFunc
+}
+
+type HealthCheckFunc = func(context.Context) error
+
+type HealthChecker struct {
+	fns []HealthCheckFunc
+}
+
+var _ HealthCheck = (*HealthChecker)(nil)
+
+// RegisterHealthCheck adds a HealthCheck to be run by the HealthChecker
+func (c *HealthChecker) RegisterHealthCheck(fn HealthCheckFunc) {
+	c.fns = append(c.fns, fn)
+}
+
+// RunHealthCheck runs all registered health checks and returns any errors combined into a single error
+func (c *HealthChecker) runHealthChecks(ctx context.Context) error {
+	var allErrors error
+	for _, fn := range c.fns {
+		if err := fn(ctx); err != nil {
+			allErrors = errors.Join(allErrors, err)
+		}
+	}
+	return allErrors
+}
+
+// HTTPHandler returns an http.Handler that performs the health checks
+func (c *HealthChecker) HTTPHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if c.runHealthChecks(r.Context()) != nil {
+			w.WriteHeader(200)
+			_, _ = w.Write([]byte("ok"))
+			return
+		}
```

Comment:
Can we make this async? i.e. periodically run the HCs in a separate goroutine and just load the result here, instead of running them on demand. The issue with the sync approach is that a misconfigured HC period can result in a DoS-like attack on the service (especially if HC is expensive, like pinging a DB connections).